lib/src/merge.rs

```diff
@@ -92,48 +98,48 @@ mod tests {
         assert_eq!(trivial_merge(&[0, 0], &[0, 0, 0]), Some(&0));
         assert_eq!(trivial_merge(&[0, 0], &[0, 0, 1]), Some(&1));
         assert_eq!(trivial_merge(&[0, 0], &[0, 1, 0]), Some(&1));
-        assert_eq!(trivial_merge(&[0, 0], &[0, 1, 1]), None);
+        assert_eq!(trivial_merge(&[0, 0], &[0, 1, 1]), Some(&1));
         assert_eq!(trivial_merge(&[0, 0], &[0, 1, 2]), None);
         assert_eq!(trivial_merge(&[0, 0], &[1, 0, 0]), Some(&1));
-        assert_eq!(trivial_merge(&[0, 0], &[1, 0, 1]), None);
+        assert_eq!(trivial_merge(&[0, 0], &[1, 0, 1]), Some(&1));
         assert_eq!(trivial_merge(&[0, 0], &[1, 0, 2]), None);
-        assert_eq!(trivial_merge(&[0, 0], &[1, 1, 0]), None);
+        assert_eq!(trivial_merge(&[0, 0], &[1, 1, 0]), Some(&1));
         assert_eq!(trivial_merge(&[0, 0], &[1, 1, 1]), Some(&1));
         assert_eq!(trivial_merge(&[0, 0], &[1, 1, 2]), None);
         assert_eq!(trivial_merge(&[0, 0], &[1, 2, 0]), None);
         assert_eq!(trivial_merge(&[0, 0], &[1, 2, 1]), None);
         assert_eq!(trivial_merge(&[0, 0], &[1, 2, 2]), None);
         assert_eq!(trivial_merge(&[0, 0], &[1, 2, 3]), None);
-        assert_eq!(trivial_merge(&[0, 1], &[0, 0, 0]), None);
+        assert_eq!(trivial_merge(&[0, 1], &[0, 0, 0]), Some(&0));
         assert_eq!(trivial_merge(&[0, 1], &[0, 0, 1]), Some(&0));
         assert_eq!(trivial_merge(&[0, 1], &[0, 0, 2]), None);
         assert_eq!(trivial_merge(&[0, 1], &[0, 1, 0]), Some(&0));
         assert_eq!(trivial_merge(&[0, 1], &[0, 1, 1]), Some(&1));
         assert_eq!(trivial_merge(&[0, 1], &[0, 1, 2]), Some(&2));
         assert_eq!(trivial_merge(&[0, 1], &[0, 2, 0]), None);
         assert_eq!(trivial_merge(&[0, 1], &[0, 2, 1]), Some(&2));
-        assert_eq!(trivial_merge(&[0, 1], &[0, 2, 2]), None);
+        assert_eq!(trivial_merge(&[0, 1], &[0, 2, 2]), Some(&2));
         assert_eq!(trivial_merge(&[0, 1], &[0, 2, 3]), None);
         assert_eq!(trivial_merge(&[0, 1], &[1, 0, 0]), Some(&0));
         assert_eq!(trivial_merge(&[0, 1], &[1, 0, 1]), Some(&1));
         assert_eq!(trivial_merge(&[0, 1], &[1, 0, 2]), Some(&2));
         assert_eq!(trivial_merge(&[0, 1], &[1, 1, 0]), Some(&1));
-        assert_eq!(trivial_merge(&[0, 1], &[1, 1, 1]), None);
+        assert_eq!(trivial_merge(&[0, 1], &[1, 1, 1]), Some(&1));
         assert_eq!(trivial_merge(&[0, 1], &[1, 1, 2]), None);
         assert_eq!(trivial_merge(&[0, 1], &[1, 2, 0]), Some(&2));
```

Comment:
This has little to do with this PR, but I find this example problematic:

```
1    2    0
 ^   ^   ^
  \ / \ /
   0   1
```

Is it OK to be this cavalier with the ordering? If we respect it, the resolution is clearly ambiguous.

Perhaps we should leave a note about it as well.

Reply:
OTOH, if we think about this as a sequence of diffs, `1 0->2 1->0`, it seems less absurd to reorder them.

Reply:
Yes, I think of these as diffs.

Let's say the user merged commits A and B with common ancestor O and resolved a conflict, resulting in commit C. The auto-merged parent state would then be `A+(B-O)` (assuming we treat the first add as the base to apply diffs onto). If the user then does `jj rebase -r C -d B -d A` (i.e. rebases the resolution onto a merge done in the opposite order), we would get the new state `B+(A-O)+(C-(A+(B-O)))` = `B+A-O+C-A-B+O`, or `removes = [O,A,B], adds=[B,A,C,O]`. If we don't allow reordering (i.e. we only allow canceling the first remove against either of the first two adds?), then we wouldn't be able to resolve this, right?

Reply:
The rest of your comments seem to depend on what we decide about ordering, so I'll await your reply to this comment before I address (or not) the other comments.

Reply:
I don't have any actionable comments. The issue with ordering comes from a previous PR. Ideally, I'd like to include an explanation of what the other options are in comments, but I'm not sure my thoughts are clear enough for that at the moment.

Here are some non-actionable thoughts.

> If we don't allow reordering (i.e. we only allow canceling the first remove against either of the first two adds?), then we wouldn't be able to resolve this, right?

I think that's right, and I think there might be some benefit in theory to not allowing that. That would allow us to preserve the distinction of "our" side and "their" side when rebasing.

Another argument for allowing reordering is something we advertise in our README. Let's say we have a commit chain `A->B->C` and we reorder B and C to get `A->C'->B'`. As a conflict, I think `B'` can be written as `A B->C A->B`. `jj` declares that `B'` should *not* be conflicted (this is what we advertise in the README) and be equivalent to `C` .  

This suggests that we might want to allow reordering that preserves the diff structure, e.g. we'd like `A B->C D->E` to be the same as `A D->E B->C` even if we differentiate it from `A B->E D->C`. In principle, this is more restrictive than the arbitrary kind of reordering we currently allow. In terms of this PR, however, this more restrictive kind of reordering seems sufficient to resolve every conflicts I was concerned about.

I think it's possible we'll come up with use-cases for preserving some (e.g. up to diff structure) or all order info, but it seems we haven't gotten there yet.

Reply:
Yeah, I have also felt that it's a bit weird to reorder `A B->C D->E` to `A B->E D->C`. Doing that means that we may produce diffs that did not appear in a diff compared to a parent. I'll work on a patch to change `Conflict::simplify()` to be more careful with how it reorders. I'll merge this as is since it doesn't change anything about ordering.

Reply:
As I said, in all of your examples, it doesn't seem to matter in the end, at least as far as `simplify` goes. This was surprising and weird to me. I suspect it would make a difference in larger conflicts, but they are not common.

Reply:
Also, I don't know if I made this clear: in my mind swapping the order of a merge is closely related to reordering `A B->C D->E` to `A B->E D->C`. That's another reason I hesitate about that in theory.

Reply:
> Also, I don't know if I made this clear: in my mind swapping the order of a merge is closely related to reordering `A B->C D->E` to `A B->E D->C`. That's another reason I hesitate about that in theory.

I can at least see how it's like reordering from `A B->C` to `C B->A`.

Reply:
Exactly. My example just has more sides. Adding an extra side to reordering from `A B->C` to `C B->A` can give you a reordering from `D E->A B->C` to `D E->C B->A`.

This might look more natural if you think of going from `∅➜A B➜C` to `∅➜D E➜A B➜C`. Note that if `D=E`, the two would be equivalent conflicts (so, you can add a side without changing the conflict, and then change one of the sides to get a non-trivial example with more sides_.